File: src/test/python/datastructures/minpriorityqueue/fibonacciheap/fibonacci_heap.py
class FibonacciHeap:
    class Node:
        def __init__(self, val):
            self.val = val
            self.left = None
            self.right = None

        def insert(self, new_node):
            if self.left is self:
                self.left = new_node
                self.right = new_node

                self.left.right = self
                self.left.left = self
            else:
                temp = self.left
                left = new_node
                left.right = self
                left.left = temp
                new_node.right = left

    def __init__(self):
        self.min = None

    def insert(self, val):
        new_node = self.Node(val)
        if not self.min:
            self.min = new_node
            self.min.left = self.min
            self.min.right = self.min
        else:
            self.min.insert(new_node)

            if val < self.min.val:
                self.min = new_node


if __name__ == "__main__":
    heap = FibonacciHeap()
    heap.insert(5)
    heap.insert(4)
    heap.insert(6)
    heap.insert(1)
    print("ended")

File: src/main/cpp/datastructures/minpriorityqueue/fibonacciheap/fibonacci_heap.cpp
//
// Created by nisha on 6/3/2020.
//

#include "fibonacci_heap.hpp"

fibonacci_heap::node::node(int _val,
                           fibonacci_heap::node *_left,
                           fibonacci_heap::node *_right,
                           fibonacci_heap::node *_parent,
                           fibonacci_heap::node *_child) :
    val(_val), left(_left), right(_right), parent(_parent), child(_child) {}

fibonacci_heap::node::~node() {

}

void fibonacci_heap::node::insert(fibonacci_heap::node *other) {
  if (left == this) {
    left = other;
    right = other;
    left->left = this;
    left->right = this;
  } else {
    node *temp = left;
    left = other;
    left->right = this;
    temp->right = left;
    left->left = temp;
  }
}

fibonacci_heap::fibonacci_heap() :
    min(nullptr) {}

fibonacci_heap::~fibonacci_heap() {
  delete min;
}

void fibonacci_heap::insert(int val) {
  if (!min) {
    min = new node(val);
    min->left = min;
    min->right = min;
  } else {
    min->insert(new node(val));
  }
}
